pennylane/gradients/jvp.py

```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
+    #     - (r_1,..,r_R)                 tensor return type and scalar parameter
+    #     - (l_1,..,l_{L_1})             scalar return type and tensor parameter
+    #     - (r_1,..,r_R, l_1,..,l_{L_1}) tensor return type and tensor parameter
+    #
+    # In this scenario, the tangent is reshaped into a one-dimensional tensor with shape (prod_l,)
+    # and the Jacobian is reshaped to have the dimensions (r_1, ... r_R, prod_l).
+    # This is followed by a tensordot contraction over the prod_l axes.
+
+    # For scenario (B), the input shapes can be in
+    #   tangent:
+    #     - (k,)                                   k scalar parameters
+    #     - [(),..,()]                             k scalar parameters
+    #     - ((),..,())                             k scalar parameters
+    #     - [(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})] k mixed scalar and tensor parameters
+    #   jac:
+    #     - ((),..,())                             scalar return type and k scalar params
+    #     - ((r_1,..,r_R),..,(r_1,..,r_R))         tensor return type and k scalar params
+    #     - ((l_1,..,l_{L_1}),..,(l_1,..,l_{L_1})) scalar return type and k mixed params
+    #     - ((r_1,..,r_R, l_1,..,l_{L_1}),..,(r_1,..,r_R, l_1,..,l_{L_k}))
+    #                                              tensor return type and k mixed params
+    #
+    # Note that the return type dimensions always remain the same, whereas the
+    # dimensions of the tensor parameters may differ.
+    #
+    # In this scenario, another case separation is used: If any of the parameters is a
+    # tensor (i.e. not a scalar), all tangent entries are reshaped into one-dimensional
+    # tensors with shapes (prod_l_i,) and then stacked to one one-dimensional tensor.
+    # If there are no tensor parameters, the tangent is just stacked and reshaped.
+    # The Jacobians are reshaped to have the dimensions (r_1, ... r_R, prod_l_i)
+    # and then are concatenated along their last (potentially mismatching) axis.
+    # This is followed by a tensordot contraction over the axes of size sum_i prod_l_i.
+
     if jac is None:
         return None
-
-    tangent = qml.math.stack(tangent)
-    jac = _convert(jac, tangent)
-
-    # Single param
-    if not isinstance(jac, tuple):
+    single_param = not isinstance(jac, tuple)
+    if (single_param and jac.shape == (0,)) or (not single_param and len(jac) == 0):
         # No trainable parameters
-        if jac.shape == (0,):
-            res = qml.math.zeros((1, 0))
-            return res
-
-        tangent = qml.math.reshape(tangent, (1,))
-
-        # No dimension e.g. expval
-        if jac.shape == ():
-            jac = qml.math.reshape(jac, (1,))
-
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
```

Comment:
Do we need to stack if we just have one parameter?

Reply:
Yes, in order to obtain a `tensor_like`. The `tangent` for example can come as a single-element list of a scalar, like `[np.array(0.5)]`, but we want to work with `tangent.shape` and `tangent.dtype` further down, and reshape it...